File: Angular/src/app/features/planes-specific/model/engine.ts
import { Validators } from '@angular/forms';
import {
  BaseDto,
  BiaFieldConfig,
  BiaFieldNumberFormat,
  BiaFieldsConfig,
  NumberMode,
  OptionDto,
  PrimeNGFiltering,
  PropType,
  VersionedDto,
} from 'biang/models';

export interface Engine extends BaseDto, VersionedDto {
  reference: string;
  manufacturer: string | null;
  nextMaintenanceDate: Date;
  lastMaintenanceDate: Date | null;
  deliveryDate: Date;
  exchangeDate: Date | null;
  syncTime: string;
  ignitionTime: string | null;
  power: number | null;
  noiseLevel: number;
  flightHours: number;
  averageFlightHours: number | null;
  fuelConsumption: number;
  averageFuelConsumption: number | null;
  originalPrice: number;
  estimatedPrice: number | null;
  isToBeMaintained: boolean;
  isHybrid: boolean | null;
  planeId: number;
  principalPart: OptionDto | null;
  installedParts: OptionDto[];
}

export const engineFieldsConfiguration: BiaFieldsConfig<Engine> = {
  columns: [
    Object.assign(new BiaFieldConfig('reference', 'engine.reference'), {
      type: PropType.String,
      isRequired: true,
    }),
    Object.assign(new BiaFieldConfig('manufacturer', 'engine.manufacturer'), {
      type: PropType.String,
    }),
    Object.assign(
      new BiaFieldConfig('nextMaintenanceDate', 'engine.nextMaintenanceDate'),
      {
        type: PropType.DateTime,
        isRequired: true,
      }
    ),
    Object.assign(
      new BiaFieldConfig('lastMaintenanceDate', 'engine.lastMaintenanceDate'),
      {
        type: PropType.DateTime,
      }
    ),
    Object.assign(new BiaFieldConfig('syncTime', 'engine.syncTime'), {
      type: PropType.TimeSecOnly,
      isRequired: true,
    }),
    Object.assign(new BiaFieldConfig('ignitionTime', 'engine.ignitionTime'), {
      type: PropType.TimeSecOnly,
    }),
    Object.assign(new BiaFieldConfig('power', 'engine.power'), {
      type: PropType.Number,
      filterMode: PrimeNGFiltering.Equals,
    }),
    Object.assign(new BiaFieldConfig('noiseLevel', 'engine.noiseLevel'), {
      type: PropType.Number,
      filterMode: PrimeNGFiltering.Equals,
      isRequired: true,
    }),
    Object.assign(new BiaFieldConfig('flightHours', 'engine.flightHours'), {
      isRequired: true,
      type: PropType.Number,
      filterMode: PrimeNGFiltering.Equals,
      displayFormat: Object.assign(new BiaFieldNumberFormat(), {
        mode: NumberMode.Decimal,
        minFractionDigits: 6,
        maxFractionDigits: 6,
      }),
      validators: [Validators.min(0)],
    }),
    Object.assign(
      new BiaFieldConfig('averageFlightHours', 'engine.averageFlightHours'),
      {
        type: PropType.Number,
        filterMode: PrimeNGFiltering.Equals,
        displayFormat: Object.assign(new BiaFieldNumberFormat(), {
          mode: NumberMode.Decimal,
          minFractionDigits: 6,
          maxFractionDigits: 6,
        }),
      }
    ),
    Object.assign(
      new BiaFieldConfig('fuelConsumption', 'engine.fuelConsumption'),
      {
        isRequired: true,
        type: PropType.Number,
        filterMode: PrimeNGFiltering.Equals,
        displayFormat: Object.assign(new BiaFieldNumberFormat(), {
          mode: NumberMode.Decimal,
          minFractionDigits: 2,
          maxFractionDigits: 2,
        }),
        validators: [Validators.min(0)],
      }
    ),
    Object.assign(
      new BiaFieldConfig(
        'averageFuelConsumption',
        'engine.averageFuelConsumption'
      ),
      {
        type: PropType.Number,
        filterMode: PrimeNGFiltering.Equals,
        displayFormat: Object.assign(new BiaFieldNumberFormat(), {
          mode: NumberMode.Decimal,
          minFractionDigits: 2,
          maxFractionDigits: 2,
        }),
        validators: [Validators.min(0), Validators.max(100)],
      }
    ),
    Object.assign(new BiaFieldConfig('originalPrice', 'engine.originalPrice'), {
      isRequired: true,
      type: PropType.Number,
      filterMode: PrimeNGFiltering.Equals,
      displayFormat: Object.assign(new BiaFieldNumberFormat(), {
        mode: NumberMode.Currency,
        minFractionDigits: 2,
        maxFractionDigits: 2,
        currency: 'EUR',
      }),
      validators: [Validators.min(0)],
    }),
    Object.assign(
      new BiaFieldConfig('estimatedPrice', 'engine.estimatedPrice'),
      {
        type: PropType.Number,
        filterMode: PrimeNGFiltering.Equals,
        displayFormat: Object.assign(new BiaFieldNumberFormat(), {
          mode: NumberMode.Currency,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          currency: 'EUR',
        }),
        validators: [Validators.min(0)],
      }
    ),
    Object.assign(new BiaFieldConfig('isHybrid', 'engine.isHybrid'), {
      isSearchable: true,
      type: PropType.Boolean,
    }),
    Object.assign(
      new BiaFieldConfig('isToBeMaintained', 'engine.isToBeMaintained'),
      {
        isRequired: true,
        isSearchable: true,
        type: PropType.Boolean,
      }
    ),
    Object.assign(new BiaFieldConfig('principalPart', 'engine.principalPart'), {
      type: PropType.OneToMany,
    }),
    Object.assign(
      new BiaFieldConfig('installedParts', 'engine.installedParts'),
      {
        type: PropType.ManyToMany,
      }
    ),
    Object.assign(new BiaFieldConfig('rowVersion', 'engine.rowVersion'), {
      isVisible: false,
      isVisibleInTable: false,
    }),
  ],
};

File: DotNet/TheBIADevCompany.BIADemo.Domain/Fleet/Entities/Engine.cs
// BIADemo only
// <copyright file="Engine.cs" company="TheBIADevCompany">
// Copyright (c) TheBIADevCompany. All rights reserved.
// </copyright>

namespace TheBIADevCompany.BIADemo.Domain.Fleet.Entities
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel.DataAnnotations.Schema;
    using BIA.Net.Core.Domain;

    /// <summary>
    /// The Engine entity.
    /// </summary>
    public class Engine : VersionedTable, IEntityFixable<int>
    {
        /// <summary>
        /// Gets or sets the id.
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// Gets or sets the Manufacturer's Serial Number.
        /// </summary>
        public string Reference { get; set; }

        /// <summary>
        /// Gets or sets the Manufacturer's name.
        /// </summary>
        public string Manufacturer { get; set; }

        /// <summary>
        /// Gets or sets the next maintenance date.
        /// </summary>
        public DateTime NextMaintenanceDate { get; set; }

        /// <summary>
        /// Gets or sets the last maintenance date.
        /// </summary>
        public DateTime? LastMaintenanceDate { get; set; }

        /// <summary>
        /// Gets or sets the delivery date.
        /// </summary>
        [Column(TypeName = "date")]
        public DateTime DeliveryDate { get; set; }

        /// <summary>
        /// Gets or sets the exchange date.
        /// </summary>
        [Column(TypeName = "date")]
        public DateTime? ExchangeDate { get; set; }

        /// <summary>
        /// Gets or sets the daily synchronisation hour.
        /// </summary>
        [Column(TypeName = "time")]
        public TimeSpan SyncTime { get; set; }

        /// <summary>
        /// Gets or sets the daily ignition hour.
        /// </summary>
        [Column(TypeName = "time")]
        public TimeSpan? IgnitionTime { get; set; }

        /// <summary>
        /// Gets or sets the power.
        /// </summary>
        public int? Power { get; set; }

        /// <summary>
        /// Gets or sets the noise level.
        /// </summary>
        public int NoiseLevel { get; set; }

        /// <summary>
        /// Gets or sets the flight hours.
        /// </summary>
        public double FlightHours { get; set; }

        /// <summary>
        /// Gets or sets the average of flight hours.
        /// </summary>
        public double? AverageFlightHours { get; set; }

        /// <summary>
        /// Gets or sets the fuel consumption.
        /// </summary>
        public float FuelConsumption { get; set; }

        /// <summary>
        /// Gets or sets the average of fuel consumption.
        /// </summary>
        public float? AverageFuelConsumption { get; set; }

        /// <summary>
        /// Gets or sets the original price.
        /// </summary>
        public decimal OriginalPrice { get; set; }

        /// <summary>
        /// Gets or sets the estimated price.
        /// </summary>
        public decimal? EstimatedPrice { get; set; }

        /// <summary>
        /// Gets or sets the Plane.
        /// </summary>
        public virtual Plane Plane { get; set; }

        /// <summary>
        /// Gets or sets the Plane id.
        /// </summary>
        public int PlaneId { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this instance is to be maintained.
        /// </summary>
        public bool IsToBeMaintained { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this instance is hybrid.
        /// </summary>
        public bool? IsHybrid { get; set; }

        /// <summary>
        /// Gets or sets the principal part.
        /// </summary>
        public Part PrincipalPart { get; set; }

        /// <summary>
        /// Gets or sets the principal part id.
        /// </summary>
        public int? PrincipalPartId { get; set; }

        /// <summary>
        /// Gets or sets the installed parts.
        /// </summary>
        public ICollection<Part> InstalledParts { get; set; }

        /// <summary>
        /// Gets or sets the installed parts. Via jointure table.
        /// </summary>
        public ICollection<EnginePart> InstalledEngineParts { get; set; }

        /// <inheritdoc/>
        public bool IsFixed { get; set; }

        /// <inheritdoc/>
        public DateTime? FixedDate { get; set; }
    }
}